NATS Manager Documentation for Contributor (#93)
kyma-project/nats-manager: docs/contributor/testing.md

1. Ensure you have the Kubecontext pointing to an existing Kubernetes cluster.

2. Execute the E2E test.

   If NATS Manager has not yet been deployed on the cluster, run:

   ```sh
   make e2e IMG=<container-registry>/nats-manager:<tag>
   ```

   Otherwise, simply run:

   ```sh
   make e2e-only
   ```

3. To adjust the log level, use the environment variable `E2E_LOG_LEVEL`.
   The accepted values are `debug`, `info`, `warn`, and `error`.

   To set the level, enter:

   ```sh
   export E2E_LOG_LEVEL="<loglevel>"
   ```

The E2E test consists of four consecutive steps. If desired, you can run them individually.

1. Ensure you have the Kubecontext pointing on an existing cluster and NATS Manager is deployed.

2. Set up the NATS CR and ensure all related resources (Pods, Services, PVCs) are set up.

   ```sh
   make e2e-setup
   ```

3. Execute a [bench test](https://docs.nats.io/using-nats/nats-tools/nats_cli/natsbench) on the NATS Server.

   ```sh
   make e2e-bench
   ```

   > **NOTE:** Running this step on slow hardware (like CI systems or k3d clusters) will result in poor performance.

>    > However, this tool is helpful to show that NATS JetStream is in an operational configuration.

Reviewer: AFAIK, `make e2e-bench` only tests the core NATS.

>    > However, this tool is helpful to show that NATS JetStream is in an operational configuration.

Author: I think it will test jetstream, due to the `--js` flag being set [here](https://github.com/kyma-project/nats-manager/blob/2e098751a14cb6a6122f7d7368f3c2eb54cf7f8d/e2e/scripts/natsbench.sh#L25)

## CI/CD

This project uses [Prow](https://docs.prow.k8s.io/docs/) and [GitHub Actions](https://docs.github.com/en/actions) as part of the development cycle.
The aim is to verify the functional correctness of the NATS Manager.

### Prow jobs

The Prow jobs that cover code of this repository reside in [their own repository](https://github.com/kyma-project/test-infra/tree/main/prow/jobs/nats-manager).
Presubmit jobs run on PRs and are marked with the prefix `pull`. Postsubmit jobs run on main after a PR was merged and carry the prefix `post`.

> For more information of each job, refer to their `description` field and the `command` and `args` field to gain insight to their execution details.

Reviewer 2: ```suggestion
For more information on execution details of each job, refer to their `description` field and the `command` and `args` fields.
```

> This information can also be accessed from your PR by inspecting the details to the job and viewing the Prow job YAML.

Reviewer 2: ```suggestion
Alternatively, you can access this information from your PR by inspecting the details to the job and viewing the Prow job `.yaml` file.
```

### GitHub Actions

GitHub Actions reside [within this module repository](https://github.com/kyma-project/nats-manager/tree/main/.github/workflows).
Pre- and postsubmit actions follow the same naming conventions as Prow jobs.

> The [Actions overview](https://github.com/kyma-project/nats-manager/actions/) allows you to see all the existing workflows and their execution details.

Reviewer 2: ```suggestion
The [Actions overview](https://github.com/kyma-project/nats-manager/actions/), shows all the existing workflows and their execution details. Here, you can also trigger a re-run of an action.
```

> A re-run of an action can be triggered here as well.

Reviewer 2: ```suggestion
```
